File: 1. Arrays And String/Question-6/Solution.java
public class Solution {
    /**
     * Solution 1: Using StringBuilder to build compressed string
     * Time Complexity: O(n) where n is the length of the string
     * Space Complexity: O(n) for the compressed string
     */
    public static String compressString(String str) {
        if (str == null || str.length() <= 1) {
            return str;
        }
        
        StringBuilder compressed = new StringBuilder();
        int countConsecutive = 1;
        
        for (int i = 1; i < str.length(); i++) {
            if (str.charAt(i) == str.charAt(i - 1)) {
                countConsecutive++;
            } else {
                compressed.append(str.charAt(i - 1));
                compressed.append(countConsecutive);
                countConsecutive = 1;
            }
        }
        
        // Add last character and its count
        compressed.append(str.charAt(str.length() - 1));
        compressed.append(countConsecutive);
        
        // Return the shorter string
        return compressed.length() < str.length() ? compressed.toString() : str;
    }
    
    /**
     * Solution 2: Calculate final length first to avoid unnecessary work
     * Time Complexity: O(n)
     * Space Complexity: O(n)
     */
    public static String compressStringOptimized(String str) {
        // Check if compression would save space
        int compressedLength = countCompression(str);
        if (compressedLength >= str.length()) {
            return str;
        }
        
        StringBuilder compressed = new StringBuilder(compressedLength); // Pre-allocate space
        int countConsecutive = 1;
        
        for (int i = 1; i < str.length(); i++) {
            if (str.charAt(i) == str.charAt(i - 1)) {
                countConsecutive++;
            } else {
                compressed.append(str.charAt(i - 1));
                compressed.append(countConsecutive);
                countConsecutive = 1;
            }
        }
        
        // Add last character and its count
        compressed.append(str.charAt(str.length() - 1));
        compressed.append(countConsecutive);
        
        return compressed.toString();
    }
    
    /**
     * Helper function to count the length of the compressed string
     */
    private static int countCompression(String str) {
        if (str == null || str.isEmpty()) {
            return 0;
        }
        
        int compressedLength = 0;
        int countConsecutive = 1;
        
        for (int i = 1; i < str.length(); i++) {
            if (str.charAt(i) == str.charAt(i - 1)) {
                countConsecutive++;
            } else {
                compressedLength += 1 + String.valueOf(countConsecutive).length();
                countConsecutive = 1;
            }
        }
        
        // Add last character and its count
        compressedLength += 1 + String.valueOf(countConsecutive).length();
        
        return compressedLength;
    }
    
    /**
     * Solution 3: Character-by-character approach with StringBuilder
     * Time Complexity: O(n)
     * Space Complexity: O(n)
     */
    public static String compressStringCharByChar(String str) {
        if (str == null || str.length() <= 1) {
            return str;
        }
        
        StringBuilder compressed = new StringBuilder();
        char[] chars = str.toCharArray();
        
        char currentChar = chars[0];
        int count = 1;
        
        for (int i = 1; i < chars.length; i++) {
            if (chars[i] == currentChar) {
                count++;
            } else {
                compressed.append(currentChar).append(count);
                currentChar = chars[i];
                count = 1;
            }
        }
        
        // Don't forget the last run of characters
        compressed.append(currentChar).append(count);
        
        return compressed.length() < str.length() ? compressed.toString() : str;
    }
    
    public static void main(String[] args) {
        String[] testCases = {
            "aabcccccaaa",  // should compress to "a2b1c5a3"
            "abcd",         // should stay "abcd"
            "aabb",         // should stay "aabb" (since "a2b2" is same length)
            "aaa",          // should compress to "a3"
            "a",            // should stay "a"
            ""              // should stay ""
        };
        
        System.out.println("Using StringBuilder approach:");
        for (String test : testCases) {
            System.out.println("\"" + test + "\" -> \"" + compressString(test) + "\"");
        }
        
        System.out.println("\nUsing optimized approach:");
        for (String test : testCases) {
            System.out.println("\"" + test + "\" -> \"" + compressStringOptimized(test) + "\"");
        }
        
        System.out.println("\nUsing char-by-char approach:");
        for (String test : testCases) {
            System.out.println("\"" + test + "\" -> \"" + compressStringCharByChar(test) + "\"");
        }
    }
}

File: 1. Arrays And String/Question-6/Solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <sstream>
using namespace std;

/**
 * Solution 1: Naive approach - Build compressed string and check length
 * Time Complexity: O(n) where n is the length of the string
 * Space Complexity: O(n) for the compressed string
 */
string compressString(const string& str) {
    if (str.length() <= 1) {
        return str;
    }
    
    string compressed;
    int countConsecutive = 1;
    for (int i = 1; i < str.length(); i++) {
        if (str[i] == str[i - 1]) {
            countConsecutive++;
        } else {
            compressed += str[i - 1] + to_string(countConsecutive);
            countConsecutive = 1;
        }
    }
    
    // Add last character and its count
    compressed += str[str.length() - 1] + to_string(countConsecutive);
    
    // Return the shorter string
    return compressed.length() < str.length() ? compressed : str;
}

/**
 * Solution 2: Calculate final length first to avoid unnecessary work
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
string compressStringOptimized(const string& str) {
    // Check if compression would save space
    int compressedLength = countCompression(str);
    if (compressedLength >= str.length()) {
        return str;
    }
    
    string compressed;
    compressed.reserve(compressedLength); // Pre-allocate space
    
    int countConsecutive = 1;
    for (int i = 1; i < str.length(); i++) {
        if (str[i] == str[i - 1]) {
            countConsecutive++;
        } else {
            compressed += str[i - 1];
            compressed += to_string(countConsecutive);
            countConsecutive = 1;
        }
    }
    
    // Add last character and its count
    compressed += str[str.length() - 1];
    compressed += to_string(countConsecutive);
    
    return compressed;
}

/**
 * Helper function to count the length of the compressed string
 */
int countCompression(const string& str) {
    if (str.empty()) {
        return 0;
    }
    
    int compressedLength = 0;
    int countConsecutive = 1;
    
    for (int i = 1; i < str.length(); i++) {
        if (str[i] == str[i - 1]) {
            countConsecutive++;
        } else {
            compressedLength += 1 + to_string(countConsecutive).length();
            countConsecutive = 1;
        }
    }
    
    // Add last character and its count
    compressedLength += 1 + to_string(countConsecutive).length();
    
    return compressedLength;
}

/**
 * Solution 3: Using a StringBuilder (stringstream in C++)
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
string compressStringWithStringBuilder(const string& str) {
    if (str.length() <= 1) {
        return str;
    }
    
    stringstream compressed;
    int countConsecutive = 1;
    
    for (int i = 1; i < str.length(); i++) {
        if (str[i] == str[i - 1]) {
            countConsecutive++;
        } else {
            compressed << str[i - 1] << countConsecutive;
            countConsecutive = 1;
        }
    }
    
    // Add last character and its count
    compressed << str[str.length() - 1] << countConsecutive;
    
    string result = compressed.str();
    return result.length() < str.length() ? result : str;
}

int main() {
    vector<string> testCases = {
        "aabcccccaaa",  // should compress to "a2b1c5a3"
        "abcd",         // should stay "abcd"
        "aabb",         // should stay "aabb" (since "a2b2" is same length)
        "aaa",          // should compress to "a3"
        "a",            // should stay "a"
        ""              // should stay ""
    };
    
    cout << "Using naive approach:" << endl;
    for (const auto& test : testCases) {
        cout << "\"" << test << "\" -> \"" << compressString(test) << "\"" << endl;
    }
    
    cout << "\nUsing optimized approach:" << endl;
    for (const auto& test : testCases) {
        cout << "\"" << test << "\" -> \"" << compressStringOptimized(test) << "\"" << endl;
    }
    
    cout << "\nUsing StringBuilder approach:" << endl;
    for (const auto& test : testCases) {
        cout << "\"" << test << "\" -> \"" << compressStringWithStringBuilder(test) << "\"" << endl;
    }
    
    return 0;
}
